diff -ru cups-1.5.0//scheduler/dirsvc.c cups//scheduler/dirsvc.c
--- cups-1.5.0//scheduler/dirsvc.c      2011-05-20 04:49:49.000000000 +0100
+++ cups//scheduler/dirsvc.c    2011-10-10 01:38:51.000000000 +0100
@@ -2399,6 +2399,9 @@
   keyvalue[i  ][0] = "pdl";
   keyvalue[i++][1] = p->pdl ? p->pdl : "application/postscript";

+  keyvalue[i  ][0] = "URF";
+  keyvalue[i++][1] = "none";
+
   if (get_auth_info_required(p, air_str, sizeof(air_str)))
   {
     keyvalue[i  ][0] = "air";
