From 5a48e2f78181bff7065e8383ecf32ef1ddad403c Mon Sep 17 00:00:00 2001
From: root <root@cent.localnet>
Date: Sun, 6 Dec 2009 18:26:45 +0000
Subject: [PATCH 05/14] boot without FPU

---
 arch/x86/kernel/cpu/bugs.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/arch/x86/kernel/cpu/bugs.c b/arch/x86/kernel/cpu/bugs.c
index 01a2652..632476d 100644
--- a/arch/x86/kernel/cpu/bugs.c
+++ b/arch/x86/kernel/cpu/bugs.c
@@ -55,8 +55,9 @@ static void __init check_fpu(void)
 	if (!boot_cpu_data.hard_math) {
 #ifndef CONFIG_MATH_EMULATION
 		printk(KERN_EMERG "No coprocessor found and no math emulation present.\n");
-		printk(KERN_EMERG "Giving up.\n");
-		for (;;) ;
+// Biff:  Without math emu we can continue here and get a boot.
+//		printk(KERN_EMERG "Giving up.\n");
+//		for (;;) ;
 #endif
 		return;
 	}
-- 
1.6.4

