diff -ru glib-2.35.1/gio/ginetaddress.c libglib/gio/ginetaddress.c
--- glib-2.35.1/gio/ginetaddress.c	2012-10-13 16:01:55.000000000 +0100
+++ libglib/gio/ginetaddress.c	2012-11-21 17:32:23.811631399 +0000
@@ -483,7 +483,11 @@
       return g_inet_address_new_from_bytes (addr, family);
     }
   else
+#ifdef DISABLE_IPV6
+    return NULL;
+#else
     return g_inet_address_new_from_bytes (in6addr_loopback.s6_addr, family);
+#endif
 }
 
 /**
@@ -510,7 +514,11 @@
       return g_inet_address_new_from_bytes (addr, family);
     }
   else
+#ifdef DISABLE_IPV6
+    return NULL;
+#else
     return g_inet_address_new_from_bytes (in6addr_any.s6_addr, family);
+#endif
 }
 
 
--- glib-2.35.1/gio/gsocket.c	2012-10-13 16:01:55.000000000 +0100
+++ libglib/gio/gsocket.c	2012-11-21 17:32:40.879630897 +0000
@@ -65,6 +65,22 @@
 #include "gcredentials.h"
 #include "glibintl.h"
 
+#ifndef IPV6_MULTICAST_HOPS
+#define IPV6_MULTICAST_HOPS 0
+#endif
+#ifndef IPV6_MULTICAST_LOOP
+#define IPV6_MULTICAST_LOOP 0
+#endif
+#ifndef IPV6_JOIN_GROUP
+#define IPV6_JOIN_GROUP 0
+#endif
+#ifndef IPV6_LEAVE_GROUP
+#define IPV6_LEAVE_GROUP 0
+#endif
+#ifndef IPV6_UNICAST_HOPS
+#define IPV6_UNICAST_HOPS 0
+#endif
+
 /**
  * SECTION:gsocket
  * @short_description: Low-level socket object
diff -ru glib-2.35.1/gio/tests/socket.c libglib/gio/tests/socket.c
--- glib-2.35.1/gio/tests/socket.c	2012-10-22 21:04:15.000000000 +0100
+++ libglib/gio/tests/socket.c	2012-11-21 17:32:29.223631244 +0000
@@ -636,6 +636,7 @@
 static void
 test_sockaddr (void)
 {
+#ifndef DISABLE_IPV6
   struct sockaddr_in6 sin6, gsin6;
   GSocketAddress *saddr;
   GInetSocketAddress *isaddr;
@@ -670,6 +671,7 @@
   g_assert_cmpint (sin6.sin6_flowinfo, ==, gsin6.sin6_flowinfo);
 
   g_object_unref (saddr);
+#endif
 }
 
 #ifdef G_OS_UNIX
