--- glib-2.23.4/gio/ginetaddress.c	2009-08-29 03:52:22.000000000 +0100
+++ libglib/gio/ginetaddress.c	2010-02-23 15:16:15.000000000 +0000
@@ -35,6 +35,12 @@
 
 #include "gioalias.h"
 
+ /* IPv6 Wildcard Address and Loopback Address defined by RFC2553 */
+#define IN6ADDR_ANY_INIT { { { 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 } } }
+#define IN6ADDR_LOOPBACK_INIT { { { 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1 } } }
+const struct in6_addr in6addr_any = IN6ADDR_ANY_INIT;
+const struct in6_addr in6addr_loopback = IN6ADDR_LOOPBACK_INIT;
+
 /**
  * SECTION:ginetaddress
  * @short_description: An IPv4/IPv6 address
