--- crosstool-ng-1.5.3/scripts/crosstool-NG.sh.in	2010-01-09 17:37:14.000000000 +0000
+++ crosstools/scripts/crosstool-NG.sh.in	2010-03-04 09:43:23.000000000 +0000
@@ -177,8 +177,10 @@
 fi

 # Check now if we can write to the destination directory:
-if [ -d "${CT_INSTALL_DIR}" ]; then
-    CT_TestAndAbort "Destination directory '${CT_INSTALL_DIR}' is not removable" ! -w $(dirname "${CT_INSTALL_DIR}")
+if [ "${CT_INSTALL_DIR_SHARED}" != "y" ]; then
+    if [ -d "${CT_INSTALL_DIR}" ]; then
+        CT_TestAndAbort "Destination directory '${CT_INSTALL_DIR}' is not removable" ! -w $(dirname "${CT_INSTALL_DIR}")
+    fi
 fi

 # Good, now grab a bit of informations on the system we're being run on,
@@ -215,8 +221,14 @@
     if [ "${CT_FORCE_EXTRACT}" = "y" -a -d "${CT_SRC_DIR}" ]; then
         CT_DoForceRmdir "${CT_SRC_DIR}"
     fi
-    if [ -d "${CT_INSTALL_DIR}" ]; then
-        CT_DoForceRmdir "${CT_INSTALL_DIR}"
+    if [ "${CT_INSTALL_DIR_SHARED}" = "y" ]; then
+        if [ -d "${CT_INSTALL_DIR}/${CT_TARGET}" ]; then
+            CT_DoForceRmdir "${CT_INSTALL_DIR}/${CT_TARGET}"
+        fi
+    else
+        if [ -d "${CT_INSTALL_DIR}" ]; then
+            CT_DoForceRmdir "${CT_INSTALL_DIR}"
+        fi
     fi
     # In case we start anew, get rid of the previously saved state directory
     if [ -d "${CT_STATE_DIR}" ]; then
--- crosstool-ng-1.5.3/config/global/paths.in	2010-01-09 17:37:05.000000000 +0000
+++ crosstools/config/global/paths.in	2010-03-04 09:57:57.000000000 +0000
@@ -60,6 +60,16 @@
 #      The reason you might also want to install elsewhere is if you are going
 #      to package your shinny new toolchain for distribution.
 
+config INSTALL_DIR_SHARED
+    bool
+    prompt "Shared install directory"
+    help
+      Normaly the install directory is deleted before the toolchain is
+      installed in it. This doesn't work if you want to use several toochain
+      sharing all the same 'bin' directory.
+      Setting this option to 'y' will only delete the current target
+      directory, leaving all the other toolchains untouched.
+
 config REMOVE_DOCS
     bool
     prompt "Remove documentation"
